### ISSUE TITLE
Support 4D tensors as torchio.Image

### ISSUE THREAD

**Issue by GFabien** [fepegar/torchio#213]
**🚀 Feature**

Make torchio accept 4D tensors (i.e. volumes with multiple channels) as input data for `torchio.Image`.

**Motivation**

I am working on a segmentation task using partial volume maps. Therefore, to take into account different tissues I need different channels as different tissues cannot be encoded to different integers. 

**Pitch**

If a 3D tensor is given, torchio's behaviour remains the same as before: the tensor is expanded to become a 4D tensor with a single channel. If a 4D tensor is given, it is directly loaded as it is. However, if there is a parameter saying that channels are at the end but should be at the beginning, dimensions are permuted so that we get channels first.

**Alternatives**

For now, the solution using torchio is to have several files and to create different attributes in the subject and concatenate these different maps on the channel dimension when I need to use them. 

**Additional context**

I'm fine with the alternative solution but I fear that it could be inefficient if there are more than 10 channels and you have to apply a spatial transform independently on all these channels compared to applying the spatial transform on the 4D tensor. I'm not sure what would be the implications on the different transforms to handle 4D tensors with multiple channels (especially intensity transforms). Maybe only `label` type image should be allowed to be 4D. 

I guess @fepegar you know better the pros and cons and why the choice was made to handle only 3D tensors in the first place.

**Comment by fepegar** [comment]
Ah, I knew this issue would be opened one day. First, let's take a look at the time it takes to resample `Colin27` with more and more channels, as I tried to do here: https://github.com/fepegar/torchio/issues/16#issuecomment-576949538

```python
import time
import torchio as tio
import numpy as np
import SimpleITK as sitk
import matplotlib.pyplot as plt

transform = sitk.ReadTransform('/tmp/affine.h5')
colin = tio.datasets.Colin27().t1.as_sitk()

times = []
for i in range(1, 21):
    print('Images:', i)
    images = i * [colin]
    composed = sitk.Compose(images)
    t1 = time.time()
    sitk.Resample(composed, transform)
    t2 = time.time()
    times.append(t2 - t1)
    [sitk.Resample(im, transform) for im in images]
    times.append(time.time() - t2)

times_composed = np.array(times[::2])
times_separate = np.array(times[1::2])

fig, ax = plt.subplots()
ax.plot(times_composed, label='Composed')
ax.plot(times_separate, label='Individual')
ax.legend()
ax.grid()
ax.set_xlabel('Number of images')
ax.set_ylabel('Time (s)')
ax.set_title('Resample applied to images of size (181, 217, 181)')
plt.show()
```

![Figure_1](https://user-images.githubusercontent.com/12688084/86227117-80d8f200-bb84-11ea-888b-1200d61e40af.png)

Resampling a 4D image is slower! I'm disappointed too.

It's true though, that images could be decomposed for resampling if needed. Actually, 4D could be useful for #168, which I haven't been able to figure out.

I do think that supporting 4D might be useful for people working with dMRI or fMRI, but I don't really know anything about DL applied to those modalities. The Brain Hackathon would've helped, as suggested by @romainVala, but I didn't manage to prepare something on time.

For now, I think the best you can do if you is open the 4D image yourself and pass the affine and each channel to different instances of `Image`, instead of the filepath.

**Comment by GFabien** [comment]
> Resampling a 4D image is slower! I know, I'm disappointed too.

I didn't expect that, this is surprising! 

> For now, I think the best you can do if you is open the 4D image yourself and pass the affine and each channel to different instances of `Image`, instead of the filepath.

In our case we have both the individual files and the stacked ones so it's OK.

**Comment by fepegar** [comment]
I'll close this for now. We can reopen if we ever consider this to be necessary.

**Comment by fepegar** [comment]
For completeness, here's a result applying a B-spline transform (on a nice DGX, not on my 2014 laptop):

![Figure_1](https://user-images.githubusercontent.com/12688084/86781838-4e346b00-c056-11ea-97db-2c926fca70ea.png)

**Comment by fepegar** [comment]
And for more completeness, here's the affine from https://github.com/fepegar/torchio/issues/213#issuecomment-652304652 using the DGX:

![Figure_1](https://user-images.githubusercontent.com/12688084/86782171-b1260200-c056-11ea-8e87-a0d1db5600bd.png)
